File: backend/cmd/mslseed/validate.go
package main

// Field-level validation of one CSV row into an admin.ServiceInput. Pure
// (no DB): the same enums/rules as admin.ServiceInput.normalize (which stays
// the authoritative server-side gate — this is an early, row-addressable
// error report so a bad seed CSV fails fast with a useful line number instead
// of a generic [data tidak lengkap...] from deep inside admin).
//
// Local literal copies of admin's pricing-mode/frequency enums: admin.go
// deliberately keeps them unexported (and never imports module0_sales, to
// avoid an import cycle — module0_sales imports admin). This cmd is a leaf
// package, so it may import both.

import (
	"fmt"
	"strings"
	"time"

	"github.com/meagrup/agencyapp/backend/internal/admin"
	"github.com/meagrup/agencyapp/backend/internal/core/money"
	"github.com/meagrup/agencyapp/backend/internal/module0_sales"
)

const (
	pricingFlat         = "flat"
	pricingMinFloor     = "min_floor"
	pricingBatchCeiling = "batch_ceiling"
	pricingPassthrough  = "passthrough"
)

var validPricingModes = map[string]bool{
	pricingFlat: true, pricingMinFloor: true, pricingBatchCeiling: true, pricingPassthrough: true,
}

var validFrequencies = map[string]bool{"Monthly": true, "One-time": true, "Campaign": true}

// errf builds a row-addressable error (line number + service_key for
// grep-ability in a 32-row CSV).
func (r Row) errf(format string, args ...any) error {
	return fmt.Errorf("baris %d (%s): "+format, append([]any{r.Line, r.ServiceKey}, args...)...)
}

// ToServiceInput validates one raw CSV row and converts it to an
// admin.ServiceInput. It is the single source of truth for "is this row
// well-formed" — both dry-run and apply call it before touching the DB.
func (r Row) ToServiceInput() (admin.ServiceInput, error) {
	name := strings.TrimSpace(r.Name)
	if name == "" {
		return admin.ServiceInput{}, r.errf("name kosong")
	}

	mode := strings.TrimSpace(r.PricingMode)
	if mode == "" {
		mode = pricingFlat
	}
	if !validPricingModes[mode] {
		return admin.ServiceInput{}, r.errf("pricing_mode tidak dikenal: %q", r.PricingMode)
	}

	freq := strings.TrimSpace(r.Frequency)
	if freq != "" && !validFrequencies[freq] {
		return admin.ServiceInput{}, r.errf("frequency tidak dikenal: %q", r.Frequency)
	}

	minQty := strings.TrimSpace(r.MinQty)
	needsMinQty := mode == pricingMinFloor || mode == pricingBatchCeiling
	if needsMinQty && minQty == "" {
		return admin.ServiceInput{}, r.errf("min_qty wajib diisi untuk pricing_mode=%s", mode)
	}
	if !needsMinQty && minQty != "" {
		return admin.ServiceInput{}, r.errf("min_qty harus kosong untuk pricing_mode=%s (hanya berlaku untuk min_floor/batch_ceiling)", mode)
	}
	if needsMinQty {
		m, err := money.Parse(minQty)
		if err != nil || m <= 0 || int64(m)%100 != 0 {
			return admin.ServiceInput{}, r.errf("min_qty bukan bilangan bulat positif: %q", r.MinQty)
		}
	}

	priceRaw := strings.TrimSpace(r.UnitPrice)
	if priceRaw == "" {
		priceRaw = "0"
	}
	price, err := money.Parse(priceRaw)
	if err != nil {
		return admin.ServiceInput{}, r.errf("unit_price tidak valid: %q", r.UnitPrice)
	}
	if mode == pricingPassthrough {
		if price < 0 {
			return admin.ServiceInput{}, r.errf("unit_price tidak boleh negatif untuk passthrough: %q", r.UnitPrice)
		}
	} else if price <= 0 {
		return admin.ServiceInput{}, r.errf("unit_price harus > 0 untuk pricing_mode=%s: %q", mode, r.UnitPrice)
	}

	applyPPN, err := parseZeroOne(r.ApplyPPN)
	if err != nil {
		return admin.ServiceInput{}, r.errf(`apply_ppn harus "0" atau "1": %q`, r.ApplyPPN)
	}

	active, err := parseTrueFalse(r.Active)
	if err != nil {
		return admin.ServiceInput{}, r.errf(`active harus "true" atau "false": %q`, r.Active)
	}

	commissionRule := strings.TrimSpace(r.CommissionRule)
	if _, err := module0_sales.ParseCommissionRule(commissionRule); err != nil {
		return admin.ServiceInput{}, r.errf("commission_rule tidak dikenal (grammar DECISIONS O14): %q", commissionRule)
	}

	effFrom := strings.TrimSpace(r.EffectiveFrom)
	if _, err := time.Parse("2006-01-02", effFrom); err != nil {
		return admin.ServiceInput{}, r.errf("effective_from harus YYYY-MM-DD: %q", r.EffectiveFrom)
	}

	return admin.ServiceInput{
		Name:           name,
		StandardPrice:  price.Decimal(),
		CommissionRule: commissionRule,
		Category:       strings.TrimSpace(r.Category),
		Unit:           strings.TrimSpace(r.Unit),
		MinQty:         minQty,
		PricingMode:    mode,
		ApplyPPN:       applyPPN,
		Frequency:      freq,
		PriceNote:      strings.TrimSpace(r.PriceNote),
		Description:    strings.TrimSpace(r.Description),
		Active:         active,
		EffectiveFrom:  effFrom,
	}, nil
}

// parseZeroOne parses the CSV boolean convention for apply_ppn ("0"/"1").
func parseZeroOne(s string) (bool, error) {
	switch strings.TrimSpace(s) {
	case "0":
		return false, nil
	case "1":
		return true, nil
	default:
		return false, fmt.Errorf("bukan 0/1: %q", s)
	}
}

// parseTrueFalse parses the CSV boolean convention for active ("true"/"false").
func parseTrueFalse(s string) (bool, error) {
	switch strings.TrimSpace(s) {
	case "true":
		return true, nil
	case "false":
		return false, nil
	default:
		return false, fmt.Errorf("bukan true/false: %q", s)
	}
}

File: apps/api/scripts/mslseed/validate.ts
/**
 * Field-level validation of one seed CSV row into an `msl.ServiceInput`.
 *
 * Port of `archive/backend-go/cmd/mslseed/validate.go`. Pure (no DB): the same enums and
 * rules as `msl.normalizeInput` — which stays the authoritative server-side gate
 * — but reported per row so a bad seed CSV fails fast with a usable line number
 * instead of a bare `[data tidak lengkap, silahkan lengkapi semua pertanyaan
 * wajib!]` thrown from deep inside the domain layer.
 *
 * The messages here are operator diagnostics for a CLI, NOT user-facing
 * validation strings: they are the verbatim Indonesian text Go's mslseed already
 * printed, so house rule #5 (exact BI `[...]` messages) is untouched — no new
 * bracketed string is invented.
 *
 * As in Go, this file may import both `msl` and `sales`: it is a leaf, while
 * `msl` deliberately never imports `sales` so the catalog admin and the pricing
 * calculator cannot form a cycle.
 */

import { money } from '@cdps/core';
import { msl, sales } from '@cdps/domain';
import type { Row } from './csv';

const PRICING_FLAT = 'flat';
const PRICING_MIN_FLOOR = 'min_floor';
const PRICING_BATCH_CEILING = 'batch_ceiling';
const PRICING_PASSTHROUGH = 'passthrough';

const VALID_PRICING_MODES = new Set([
  PRICING_FLAT, PRICING_MIN_FLOOR, PRICING_BATCH_CEILING, PRICING_PASSTHROUGH,
]);

const VALID_FREQUENCIES = new Set(['Monthly', 'One-time', 'Campaign']);

/** A row-addressable seed validation failure (line number + service_key). */
export class SeedRowError extends Error {
  readonly line: number;
  readonly serviceKey: string;

  constructor(row: Row, detail: string) {
    super(`baris ${row.line} (${row.serviceKey}): ${detail}`);
    this.name = 'SeedRowError';
    this.line = row.line;
    this.serviceKey = row.serviceKey;
  }
}

/** q renders a value the way Go's %q verb did, for identical CLI output. */
function q(s: string): string {
  return JSON.stringify(s);
}

/**
 * rowToServiceInput validates one raw CSV row and converts it to an
 * `msl.ServiceInput`. It is the single source of truth for "is this row
 * well-formed" — both dry-run and apply call it before touching the DB.
 */
export function rowToServiceInput(row: Row): msl.ServiceInput {
  const name = row.name.trim();
  if (name === '') {
    throw new SeedRowError(row, 'name kosong');
  }

  let mode = row.pricingMode.trim();
  if (mode === '') {
    mode = PRICING_FLAT;
  }
  if (!VALID_PRICING_MODES.has(mode)) {
    throw new SeedRowError(row, `pricing_mode tidak dikenal: ${q(row.pricingMode)}`);
  }

  const frequency = row.frequency.trim();
  if (frequency !== '' && !VALID_FREQUENCIES.has(frequency)) {
    throw new SeedRowError(row, `frequency tidak dikenal: ${q(row.frequency)}`);
  }

  const minQty = row.minQty.trim();
  const needsMinQty = mode === PRICING_MIN_FLOOR || mode === PRICING_BATCH_CEILING;
  if (needsMinQty && minQty === '') {
    throw new SeedRowError(row, `min_qty wajib diisi untuk pricing_mode=${mode}`);
  }
  if (!needsMinQty && minQty !== '') {
    throw new SeedRowError(
      row,
      `min_qty harus kosong untuk pricing_mode=${mode} (hanya berlaku untuk min_floor/batch_ceiling)`,
    );
  }
  if (needsMinQty) {
    let m: money.Money;
    try {
      m = money.parse(minQty);
    } catch {
      throw new SeedRowError(row, `min_qty bukan bilangan bulat positif: ${q(row.minQty)}`);
    }
    if (m <= 0n || m % 100n !== 0n) {
      throw new SeedRowError(row, `min_qty bukan bilangan bulat positif: ${q(row.minQty)}`);
    }
  }

  let priceRaw = row.unitPrice.trim();
  if (priceRaw === '') {
    priceRaw = '0';
  }
  let price: money.Money;
  try {
    price = money.parse(priceRaw);
  } catch {
    throw new SeedRowError(row, `unit_price tidak valid: ${q(row.unitPrice)}`);
  }
  if (mode === PRICING_PASSTHROUGH) {
    if (price < 0n) {
      throw new SeedRowError(row, `unit_price tidak boleh negatif untuk passthrough: ${q(row.unitPrice)}`);
    }
  } else if (price <= 0n) {
    throw new SeedRowError(row, `unit_price harus > 0 untuk pricing_mode=${mode}: ${q(row.unitPrice)}`);
  }

  const applyPPN = parseZeroOne(row.applyPPN);
  if (applyPPN === null) {
    throw new SeedRowError(row, `apply_ppn harus "0" atau "1": ${q(row.applyPPN)}`);
  }

  const active = parseTrueFalse(row.active);
  if (active === null) {
    throw new SeedRowError(row, `active harus "true" atau "false": ${q(row.active)}`);
  }

  // Grammar gate (DECISIONS O14): only "<N>% of standard price" or "flat Rp <N>".
  const commissionRule = row.commissionRule.trim();
  try {
    sales.parseCommissionRule(commissionRule);
  } catch {
    throw new SeedRowError(row, `commission_rule tidak dikenal (grammar DECISIONS O14): ${q(commissionRule)}`);
  }

  const effectiveFrom = row.effectiveFrom.trim();
  if (!isCalendarDate(effectiveFrom)) {
    throw new SeedRowError(row, `effective_from harus YYYY-MM-DD: ${q(row.effectiveFrom)}`);
  }

  return {
    name,
    standardPrice: money.decimal(price),
    commissionRule,
    category: row.category.trim(),
    unit: row.unit.trim(),
    minQty,
    pricingMode: mode,
    applyPPN,
    frequency,
    priceNote: row.priceNote.trim(),
    description: row.description.trim(),
    active,
    effectiveFrom,
  };
}

/** parseZeroOne parses the CSV boolean convention for apply_ppn ("0"/"1"). */
function parseZeroOne(s: string): boolean | null {
  switch (s.trim()) {
    case '0': return false;
    case '1': return true;
    default: return null;
  }
}

/** parseTrueFalse parses the CSV boolean convention for active ("true"/"false"). */
function parseTrueFalse(s: string): boolean | null {
  switch (s.trim()) {
    case 'true': return true;
    case 'false': return false;
    default: return null;
  }
}

/**
 * isCalendarDate reports whether s is a real YYYY-MM-DD calendar date — the
 * equivalent of Go's `time.Parse("2006-01-02", s)`, which rejects both a wrong
 * shape and an out-of-range day such as 2026-02-30. A plain regex would accept
 * the latter, so the parsed date is round-tripped back to a string.
 */
function isCalendarDate(s: string): boolean {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s)) {
    return false;
  }
  const d = new Date(`${s}T00:00:00Z`);
  return !Number.isNaN(d.getTime()) && d.toISOString().slice(0, 10) === s;
}
